theano/contrib/mkl/basic_ops.py

```diff
@@ -0,0 +1,472 @@
+import theano.tensor as T
+from theano.gof import Apply, Op
+from theano.tensor.blas import ldflags
+from theano.tensor.nnet.abstract_conv import get_conv_output_shape
+from theano.contrib.mkl.mkl_helper import header_text
+from theano.contrib.mkl.mkl_type import MKLNdarrayType
+
+
+class MKLOp(Op):
+    def c_lib_dirs(self):
+        return ldflags(libs=False, libs_dir=True)
+
+    def c_libraries(self):
+        return ldflags()
+
+    def c_compile_args(self):
+        compile_args = ldflags(libs=False, flags=True)
+        compile_args += super(MKLOp, self).c_compile_args()
+        return compile_args
+
+    def c_support_code(self):
+        ccode = header_text()
+        ccode += """
+        #define DIMENSION  4
+
+        #define CHECK_ERR(f, err) \\
+            do { \\
+                (err) = (f); \\
+                if ((err) != E_SUCCESS) { \\
+                    printf("Error in file [%s:%d], err code (%d)", \\
+                           __FILE__, __LINE__, err); \\
+                    exit(1); \\
+                } \\
+            } while(0)
+        """
+
+        return ccode
+
+
+class BaseConvertOp(MKLOp):
+    def c_support_code_struct(self, node, name):
+        ccode = """
+        dnnError_t err;
+        int first_run;
+        void *internal_buf;
+        void *user_buf;
+        dnnLayout_t layout_internal;
+        dnnLayout_t layout_user;
+        dnnPrimitive_t to_internal;
+        dnnPrimitive_t from_internal;
+        dnnPrimitive_t primitive;
+        void *convert_resources[dnnResourceNumber];
+        size_t bottomSize[DIMENSION];
+        size_t bottomStride[DIMENSION];
+        """
+        return ccode
+
+    def c_init_code_struct(self, node, name, sub):
+        ccode = """
+        first_run = 1;
+        internal_buf = NULL;
+        user_buf = NULL;
+        layout_internal = NULL;
+        layout_user = NULL;
+        to_internal = NULL;
+        from_internal = NULL;
+        primitive = NULL;
+        """
+        return ccode
+
+    def c_code_cache_version(self):
+        return (1, 0)
+
+
+class I2U(BaseConvertOp):
+    __props__ = ()
+
+    def make_node(self, x):
+        assert isinstance(x.type, MKLNdarrayType)
+        return Apply(self, [x], [T.TensorType(broadcastable=x.broadcastable, dtype=x.dtype)()])
+
+    def grad(self, inp, grads):
+        x, = inp
+        gz, = grads
+
+        return [I2UGrad()(x, gz)]
```

Comment:
It is a bad practice to keep x just for its shape. This will force it to be kept for the gradient instead of being freed. This will cause higher memory peak usage.

I think you need to create only 2 class: U2I and I2U. Each are the grad of the other.

Can you infer the shape from an MKLNdarray of the corresponding Ndarray? If not, you could make one of the op take 2 inputs, the data and the shape. This will only kept the shape in the graph and the data itself.